framework/include/utils/IndexableProperty.h

```diff
@@ -0,0 +1,131 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseTypes.h"
+#include <vector>
+
+template <typename T, bool is_ad>
+class IndexableProperty
+{
+public:
+  static InputParameters validParams();
+
+  IndexableProperty(T * host,
+                    const std::string & property_param = "property",
+                    const std::string & component_param = "component");
+
+  GenericReal<is_ad> operator[](int i) const;
+  operator bool() const;
+
+protected:
+  template <typename P>
+  void getPropertyHelper(const GenericMaterialProperty<P, is_ad> *& pointer,
+                         unsigned int components);
+
+  const GenericMaterialProperty<Real, is_ad> * _property_real;
+  const GenericMaterialProperty<std::vector<Real>, is_ad> * _property_std_vector;
+  const GenericMaterialProperty<RealVectorValue, is_ad> * _property_real_vector_value;
+  const GenericMaterialProperty<RankTwoTensor, is_ad> * _property_rank_two_tensor;
+  const GenericMaterialProperty<RankThreeTensor, is_ad> * _property_rank_three_tensor;
+  const GenericMaterialProperty<RankFourTensor, is_ad> * _property_rank_four_tensor;
+
+  T * _host;
```

Comment:
does it need the host only for starting with the right valid params? Couldn't we keep that part in the host validParams?
I feel like this could be a lot less instantiations of the template

Reply:
Needed also for getParam and getMateriaProperty below